pkg/action/executor/horizontal_scaler.go

```diff
@@ -0,0 +1,308 @@
+package executor
+
+import (
+	"errors"
+	"fmt"
+	"strings"
+
+	"k8s.io/kubernetes/pkg/api"
+	"k8s.io/kubernetes/pkg/apis/extensions"
+	client "k8s.io/kubernetes/pkg/client/unversioned"
+
+	"github.com/vmturbo/kubeturbo/pkg/action/turboaction"
+	"github.com/vmturbo/kubeturbo/pkg/action/util"
+	"github.com/vmturbo/kubeturbo/pkg/discovery/probe"
+	turboscheduler "github.com/vmturbo/kubeturbo/pkg/scheduler"
+	"github.com/vmturbo/kubeturbo/pkg/turbostore"
+
+	"github.com/turbonomic/turbo-go-sdk/pkg/proto"
+
+	"github.com/golang/glog"
+)
+
+type HorizontalScaler struct {
+	kubeClient *client.Client
+	broker     turbostore.Broker
+	scheduler  *turboscheduler.TurboScheduler
+}
+
+func NewHorizontalScaler(client *client.Client, broker turbostore.Broker, scheduler *turboscheduler.TurboScheduler) *HorizontalScaler {
+	return &HorizontalScaler{
+		kubeClient: client,
+		broker:     broker,
+		scheduler:  scheduler,
+	}
+}
+
+func (h *HorizontalScaler) Execute(actionItem *proto.ActionItemDTO) (*turboaction.TurboAction, error) {
+	if actionItem == nil {
+		return nil, errors.New("ActionItem passed in is nil")
+	}
+	action, err := h.buildPendingScalingTurboAction(actionItem)
+	if err != nil {
+		return nil, err
+	}
+
+	return h.horizontalScale(action)
+}
+
+func (h *HorizontalScaler) buildPendingScalingTurboAction(actionItem *proto.ActionItemDTO) (*turboaction.TurboAction,
+	error) {
+	targetSE := actionItem.GetTargetSE()
+	targetEntityType := targetSE.GetEntityType()
+	if targetEntityType != proto.EntityDTO_CONTAINER_POD || targetEntityType != proto.EntityDTO_APPLICATION {
+		return nil, errors.New("The target service entity for scaling action is " +
+			"neither a Pod nor an Application.")
+	}
+
+	providerPod, err := h.getProviderPod(actionItem)
+	if err != nil {
+		return nil, fmt.Errorf("Try to scaling %s, but cannot find a pod related to it in the cluster: %s",
+			targetSE.GetId(), err)
+	}
+
+	targetObject := &turboaction.TargetObject{
+		TargetObjectUID:       string(providerPod.UID),
+		TargetObjectNamespace: providerPod.Namespace,
+		TargetObjectName:      providerPod.Name,
+		TargetObjectType:      turboaction.TypePod,
+	}
+
+	var parentObjRef *turboaction.ParentObjectRef
+	parentRefObject, _ := probe.FindParentReferenceObject(providerPod)
+	if parentRefObject != nil {
+		parentObjRef = &turboaction.ParentObjectRef{
+			ParentObjectUID:       string(parentRefObject.UID),
+			ParentObjectNamespace: parentRefObject.Namespace,
+			ParentObjectName:      parentRefObject.Name,
+			ParentObjectType:      parentRefObject.Kind,
+		}
+	} else {
+		return nil, errors.New("Cannot automatically scale")
```

Comment:
Change it to "Cannot perform auto-scale, please make sure the pod is connected to RC"